internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,666 @@
+//
+// TODO:
+//     Do not derive PartialEq implementation.
+//     Instead provide a custom implementation
+//     that instantiates both polytypes with the
+//     same type variables.
+//
+//     Note this requires a substitution, so remove
+//     this derivation once substitutions are defined.
+//
+#[derive(Debug, Clone, PartialEq)]
+pub struct PolyType {
+    pub free: Vec<Tvar>,
+    pub bnds: Option<HashMap<Tvar, Kind>>,
+    pub expr: MonoType,
+}
+
+impl fmt::Display for PolyType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self.bnds {
+            Some(bnds) => {
+                let mut bounds = Vec::new();
+                for tv in &self.free {
+                    if let Some(kind) = bnds.get(tv) {
+                        bounds.push(BoundTvar {
+                            tv: *tv,
+                            kind: *kind,
+                        })
+                    }
+                }
+                write!(
+                    f,
+                    "forall [{}] where {} {}",
+                    List { values: &self.free },
+                    List { values: &bounds },
+                    self.expr,
+                )
+            }
+            None => write!(f, "forall [{}] {}", List { values: &self.free }, self.expr),
+        }
+    }
+}
+
+// Kind represents a class or family of types
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Kind {
+    Add,
+    Div,
+    Cmp,
+    Nul,
+}
+
+impl fmt::Display for Kind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Kind::Add => f.write_str("Addable"),
+            Kind::Div => f.write_str("Divisible"),
+            Kind::Cmp => f.write_str("Comparable"),
+            Kind::Nul => f.write_str("Nullable"),
+        }
+    }
+}
+
+// MonoType represents a specific named type
+#[derive(Debug, Clone, PartialEq)]
+pub enum MonoType {
+    Bool,
+    Int,
+    Uint,
+    Flt,
+    Str,
+    Dur,
+    Time,
+    Rgx,
+    Var(Tvar),
+    Arr(Box<Array>),
+    Rec(Box<Row>),
+    Fun(Box<Function>),
+}
+
+impl fmt::Display for MonoType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            MonoType::Bool => f.write_str("bool"),
+            MonoType::Int => f.write_str("int"),
+            MonoType::Uint => f.write_str("uint"),
+            MonoType::Flt => f.write_str("flt"),
+            MonoType::Str => f.write_str("str"),
+            MonoType::Dur => f.write_str("dur"),
+            MonoType::Time => f.write_str("time"),
+            MonoType::Rgx => f.write_str("rgx"),
+            MonoType::Var(var) => var.fmt(f),
+            MonoType::Arr(arr) => arr.fmt(f),
+            MonoType::Rec(obj) => obj.fmt(f),
+            MonoType::Fun(fun) => fun.fmt(f),
+        }
+    }
+}
+
+// Tvar stands for type variable.
+// A type variable holds an unknown type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct Tvar(i64);
```

Comment:
Why does this type need `Hash`?

Reply:
Substitutions will be implemented as hash maps where the keys are type variables.